File: leet/0113-path-sum-iii/path-sum-iii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) [][]int {
    if root == nil {return [][]int{}}
    
    left := sum - root.Val
    r := make([]int, 1)
    r[0] = root.Val
    
    if left == 0 && root.Left == nil && root.Right == nil{
        result := make([][]int, 1)
        result[0] = r
        return result
    }
    
    leftResult := pathSum(root.Left, left)
    rightResult := pathSum(root.Right, left)
    
    allResult := append(leftResult, rightResult...)
    
    for i := 0; i < len(allResult); i++ {
        allResult[i] = append(r, allResult[i]...)
    }
    
    return allResult
}

File: leet/0437-path-sum-III/path-sum-III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if root is None: 
            return 0
        
        rs = self.findSum(root, sum)
        
        rs += self.pathSum(root.left, sum)
        rs += self.pathSum(root.right, sum)
        
        return rs
    
    def findSum(self, root: TreeNode, sum:int) -> int:
        if root is None: 
            return 0
        
        diff = sum - root.val
        rs = 0
        
        if diff == 0:
            rs += 1
        rs += self.findSum(root.left, diff)
        rs += self.findSum(root.right, diff)
        
        return rs
